ibc-clients/ics07-tendermint/src/client_state/validation.rs

```diff
@@ -190,3 +200,52 @@ where
 
     Ok(Status::Active)
 }
+
+/// Check that the subject and substitute client states match as part of
+/// the client recovery validation step.
+///
+/// The subject and substitute client states match if all their respective
+/// client state parameters match except for frozen height, latest height,
+/// trusting period, and chain ID.
+pub fn check_substitute<V>(
+    subject_client_state: &ClientStateType,
+    _ctx: &V,
+    substitute_client_state: Any,
+) -> Result<(), ClientError>
+where
+    V: TmValidationContext,
+    // V::ClientStateRef: TryInto<ClientStateType, Error = ClientError>,
+    V::ConsensusStateRef: ConsensusStateConverter,
+{
+    let subject = ClientStateType {
+        latest_height: Height::new(0, 1).expect("Panic while creating a Height { 0, 1 }"),
+        frozen_height: None,
+        trusting_period: Duration::ZERO,
+        chain_id: ChainId::new("").expect("Panic while creating an empty chain ID"),
+        allow_update: AllowUpdate {
+            after_expiry: true,
+            after_misbehaviour: true,
+        },
```

Comment:
I love this trick. But I would prefer an optimized code.

Reply:
What do you mean by "an optimized code"?

Reply:
However, now I see that if new fields are introduced in the future, this is error-prone, as we may forget to add the new field here. (I gave the same logic for the `Status` match in a separate comment.)

Reply:
Sorry, it's a bit unclear from your comment about this being error-prone. Are you saying we should make the change you suggested or not?

Reply:
Sorry for the confusion. I meant to say that my suggestion is error-prone, as it may miss newly added fields in the future.

Your changes implement the logic of checking equality of `ClientStateType` fields - except `latest_height`, `frozen_height`, `trusting_period`, `chain_id`, and  `allow_update` - which is more robust IMO.

Feel free to ignore my suggestion now.

Reply:
cc-ing @Farhad-Shabani for your opinion, just in case. :slightly_smiling_face:

Reply:
I came up with the following.

```rs
    let ClientStateType {
        latest_height: _,
        frozen_height: _,
        trusting_period: _,
        chain_id: _,
        allow_update: _,
        trust_level: subject_trust_level,
        unbonding_period: subject_unbonding_period,
        max_clock_drift: subject_max_clock_drift,
        proof_specs: subject_proof_specs,
        upgrade_path: subject_upgrade_path,
    } = subject_client_state.clone();

    let substitute_client_state = ClientStateType::try_from(substitute_client_state)?;

    let ClientStateType {
        latest_height: _,
        frozen_height: _,
        trusting_period: _,
        chain_id: _,
        allow_update: _,
        trust_level: substitute_trust_level,
        unbonding_period: substitute_unbonding_period,
        max_clock_drift: substitute_max_clock_drift,
        proof_specs: substitute_proof_specs,
        upgrade_path: substitute_upgrade_path,
    } = substitute_client_state;

    (subject_trust_level == substitute_trust_level
        && subject_unbonding_period == substitute_unbonding_period
        && subject_max_clock_drift == substitute_max_clock_drift
        && subject_proof_specs == substitute_proof_specs
        && subject_upgrade_path == substitute_upgrade_path)
        .then_some(())
        .ok_or(ClientError::ClientRecoveryStateMismatch)
```

Reply:
Nice, that's a neat suggestion 🙂

Reply:
[8fd28fa](https://github.com/cosmos/ibc-rs/pull/1127/commits/8fd28faef057e4d1afae2041c1671ea43858d91d)